caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/routes/RoutingCacheProviderWrapper.java

```diff
@@ -0,0 +1,241 @@
+package org.commonjava.maven.galley.cache.routes;
+
+import org.commonjava.maven.galley.model.ConcreteResource;
+import org.commonjava.maven.galley.model.Location;
+import org.commonjava.maven.galley.model.Transfer;
+import org.commonjava.maven.galley.spi.cache.CacheProvider;
+
+import javax.enterprise.inject.Alternative;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+
+@Alternative
+public class RoutingCacheProviderWrapper
+        implements CacheProvider
+{
+    private final CacheProvider safe;
+
+    private final CacheProvider disposable;
+
+    private final RouteSelector selector;
+
+    public RoutingCacheProviderWrapper( final RouteSelector selector, final CacheProvider disposable,
+                                        final CacheProvider safe )
+    {
+        this.disposable = disposable;
+        this.safe = safe;
+        this.selector = selector;
+    }
+
+    public CacheProvider getRoutedProvider( ConcreteResource resource )
```

Comment:
I think this might leak some implementation details to the outside world that they don't need to know about. Is it possible to reduce its scope a bit, maybe to protected?